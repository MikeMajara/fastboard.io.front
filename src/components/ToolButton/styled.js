import styled, { css } from '@xstyled/styled-components';
import { getColor, getTransition, variant } from '@xstyled/system';

export const SWrapper = styled.div`
  display: flex;
  width: toolButton;
  height: toolButton;
  box-sizing: border-box;
  align-items: center;
  justify-content: center;
  border-style: default;
  border-color: ${({ color, selected }) => ((color && selected) ? color : getColor('border'))};
  margin: auto sm;
  background-color: bg;
  border-radius: rounded;
  transition: box-shadow ${getTransition('default')};

  &:hover {
    box-shadow: default;
    cursor: pointer;
  }
  &:active {
    background-color: bgActive;
  }

  ${variant({
    default  : false,
    prop     : 'isLoading',
    variants : {
      true : css`
        pointer-events: none;
      `,
      false : css`
        cursor: pointer;
        &:hover {
          box-shadow: default;
        }
      `,
    },
  })}

  ${variant({
    default  : false,
    prop     : 'isSelected',
    variants : {
      true  : css`border-width: xlg;`,
      false : css`border-width: default;`,
    },
  })}
`;

export const SIcon = styled.span`
  & > svg {
    width: 18px;
    height: 18px;
    color: ${({ color }) => color};
  }
`;
