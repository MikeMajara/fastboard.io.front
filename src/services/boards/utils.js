/**
 * Send boards event trough the network.
 *
 * @param {String} eventName Event key name.
 * @param {Object} data Data associated to the event.
 * @param {Function} ask Callback for server event acknowledgement.
 */
function send(eventName, data, ack) {
  this.socket.emit(eventName, data, ack);
}

/**
 * Check if the service is initialized.
 *
 * @return {Boolean} True if the service is initialized; false otherwise.
 */
function isInit() {
  return this.isInit;
}

export {
  isInit,
  send,
};
