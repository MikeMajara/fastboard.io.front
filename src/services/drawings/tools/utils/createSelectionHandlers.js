import { Group, Shape } from 'paper';

import { Log } from '#utils';
import { cursorTypes, selectionColorCode } from '#constants';
import store, { setSelectorCursorHover } from '#store';

import { bounds } from '../selector/constants';

const boundsArray = Object.values(bounds);
const bounds2cursorType = Object.freeze({
  [bounds.topLeft]     : cursorTypes.rotating,
  [bounds.topRight]    : cursorTypes.menu,
  [bounds.bottomLeft]  : cursorTypes.resizingSW,
  [bounds.bottomRight] : cursorTypes.resizingSE,
});

/**
 * Add selection handlers to the selected item.
 *
 * @param {Object} selectedItem Item to be added selection handlers to.
 * @param {Object} selectionLayer Layer that will contain the selection handlers.
 *
 * @returns {Object} Item - group with 4 handler items - from the selection handler layer.
 */
export default function createSelectionHandlers(selectedItem, selectionLayer) {
  Log.debug('Service : Drawings : Tools : Utils : createSelectionHandlers', { selectedItem });

  // Set selected item to show its selection state
  selectedItem.set({
    selected      : true,
    selectedColor : selectionColorCode,
    onMouseEnter  : () => store.dispatch(setSelectorCursorHover(cursorTypes.dragging)),
    onMouseLeave  : () => store.dispatch(setSelectorCursorHover()),
  });

  // Create bound handlers on the selection layer
  boundsArray.forEach((bound) => {
    selectionLayer.addChild(new Shape.Circle({
      name         : bound,
      center       : selectedItem.internalBounds[bound],
      radius       : 6,
      fillColor    : selectionColorCode,
      onMouseEnter : () => store.dispatch(setSelectorCursorHover(bounds2cursorType[bound])),
      onMouseLeave : () => store.dispatch(setSelectorCursorHover()),
    }));
  });

  // Create a group with the handlers to manipulate all of them at once
  selectionLayer.addChild(
    new Group({
      name     : 'selectionHandlers', // @todo to constants
      children : boundsArray.map(
        (bound) => selectionLayer.children[bound],
      ),
    }),
  );

  // Reproduce the selected item transformations on the handler items group
  selectionLayer.children.selectionHandlers.transform(selectedItem.matrix);

  // Except the scaling of the handler items themselves, which retain the original size
  selectionLayer.children.selectionHandlers.children.forEach((handlerItem) => {
    handlerItem.set({ scaling : 1 });
  });

  return selectionLayer.children.selectionHandlers;
}
