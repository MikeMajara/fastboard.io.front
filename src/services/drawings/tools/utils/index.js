export { default as createSelectionHandlers } from './createSelectionHandlers';
export { default as removeSelectionHandlers } from './removeSelectionHandlers';
export { default as resizeSelectionHandlers } from './resizeSelectionHandlers';
export { default as rotateSelectionHandlers } from './rotateSelectionHandlers';
export { default as translateSelectionHandlers } from './translateSelectionHandlers';
