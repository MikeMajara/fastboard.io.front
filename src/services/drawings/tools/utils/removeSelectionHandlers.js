import { Log } from '#utils';

/**
 * Remove selection handlers.
 *
 * @param {Object} selectedItem Item to be removed selection handlers from.
 * @param {Object} selectionLayer Layer that contains the selection handlers.
 */
export default function removeSelectionHandlers(selectedItem, selectionLayer) {
  Log.debug('Service : Drawings : Tools : Utils : removeSelectionHandlers');

  selectedItem.set({ selected : false });
  selectionLayer.removeChildren();
}
