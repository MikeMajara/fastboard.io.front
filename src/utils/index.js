export { default as Log } from './logger';
export { default as compressImage } from './compressImage';
export { default as formatDate } from './formatDate';
export { default as getScaleFactor } from './getScaleFactor';
export { default as point2net } from './point2net';
export { default as setPreventUnload } from './setPreventUnload';
export { default as throttle } from './throttle';
export { default as timeoutPromise } from './timeoutPromise';
export { default as toClipboard } from './toClipboard';
export { default as triggerDownload } from './triggerDownload';
export { default as triggerUpload } from './triggerUpload';
export { default as uploadImage } from './uploadImage';
export { default as validBoardId } from './validBoardId';
export { default as validBoardUrl } from './validBoardUrl';
export { default as extractBoardId } from './extractBoardId';
export { default as TimeoutError } from './TimeoutError';
export { default as BoardError } from './BoardError';
