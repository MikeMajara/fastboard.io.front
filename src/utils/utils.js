/**
 * Throttles a function call.
 *
 * @param {Function} callback Function to be throttled.
 * @param {Integer} delay Throttle delay in milliseconds.
 *
 * @return {Function} Function that throttles the execution of the callback.
 */
export const throttle = (callback, delay = 0) => {
  let previousTime = new Date().getTime();

  return function throttledCallback(args) {
    const currentTime = new Date().getTime();

    if ((currentTime - previousTime) >= delay) {
      previousTime = currentTime;
      callback(...args);
    }
  };
};

/**
 * Creates a Promise with a rejection timeout.
 *
 * @param {Function} callback Function for the promise. Receives res and rej functions.
 * @param {Integer} timeout Timeout for the promise to be rejected in milliseconds.
 *
 * @return {Promise} Promise with rejection timeout.
 */
export const timeoutPromise = (callback, timeout) => new Promise(
  (res, rej) => {
    const timeoutId = setTimeout(() => rej(new Error()), timeout);

    callback(
      (...args) => {
        clearTimeout(timeoutId);
        res(...args);
      },
      (...args) => {
        clearTimeout(timeoutId);
        rej(new Error(...args));
      },
    );
  },
);
