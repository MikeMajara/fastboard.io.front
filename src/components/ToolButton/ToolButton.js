import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';

import { drawingColors, drawingColorCodes } from '#constants';

import Spinner from '../Spinner';
import { SWrapper, SIcon } from './styled';

const testId = 'toolbutton-component';

const ToolButton = forwardRef(({
  icon, color, isLoading, isSelected, onClick, children,
}, ref) => (
  <SWrapper
    ref={ref}
    color={drawingColorCodes[color]}
    onClick={onClick}
    isLoading={isLoading}
    isSelected={isSelected}
    data-testid={testId}
  >
    {isLoading ? <Spinner size="md" dark /> : (
      <>
        <SIcon color={drawingColorCodes[color]}>
          {icon}
        </SIcon>
        {children}
      </>
    )}
  </SWrapper>
));

ToolButton.defaultProps = {
  children   : undefined,
  color      : undefined,
  isLoading  : false,
  isSelected : false,
};

ToolButton.propTypes = {
  children   : PropTypes.node,
  color      : PropTypes.oneOf(Object.values(drawingColors)),
  icon       : PropTypes.node.isRequired,
  onClick    : PropTypes.func.isRequired,
  isLoading  : PropTypes.bool,
  isSelected : PropTypes.bool,
};

export default ToolButton;
