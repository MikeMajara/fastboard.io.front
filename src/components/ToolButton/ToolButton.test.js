import React from 'react';
import { fireEvent } from '@testing-library/react';

import { render } from '#devTools/tests/utils';

import ToolButton from './ToolButton';

describe('Component : ToolButton', () => {
  test('Is visible and clickable', () => {
    const onClickFn = jest.fn();
    const { getByTestId } = render(
      <ToolButton
        onClick={onClickFn}
        icon={<span />}
      />,
    );

    expect(getByTestId('toolbutton-component')).toBeInTheDocument();
    fireEvent.click(getByTestId('toolbutton-component'));
    expect(onClickFn).toHaveBeenCalledTimes(1);
  });

  test('Children are visible', () => {
    const onClickFn = jest.fn();
    const childrenTestId = 'toolbutton-children';
    const children = <span data-testid={childrenTestId} />;
    const { getByTestId } = render(
      <ToolButton
        onClick={onClickFn}
        icon={<span />}
      >
        {children}
      </ToolButton>,
    );

    expect(getByTestId(childrenTestId)).toBeInTheDocument();
  });

  test('Snapshot', () => {
    const onClickFn = jest.fn();
    const { asFragment } = render(
      <ToolButton
        onClick={onClickFn}
        icon={<span />}
      />,
    );

    const componentRender = asFragment();
    expect(componentRender).toMatchSnapshot();
  });
});
