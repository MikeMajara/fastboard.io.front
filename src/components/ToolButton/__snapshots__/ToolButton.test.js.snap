// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component : ToolButton Snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-qQYBZ iGSSSO"
    data-testid="toolbutton-component"
  >
    <span
      class="sc-qZtCU kwFhvv"
    >
      <span />
    </span>
  </div>
</DocumentFragment>
`;
